### PR TITLE
fix: explain pipeline bug of update/delete

### LINKED ISSUE #12597: bug: explain pipeline panic of update/delete
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

[0c5dd54](https://github.com/datafuselabs/databend/commit/0c5dd54b47a2ab94e213d9dd517d7f52d3598071)

### What's Wrong?

mysql> explain pipeline update t1 set a = 1;
ERROR 1105 (HY000): TokioError. Code: 1068, Text = Cannot join handle from context's runtime, cause: task 262 panicked.
mysql> explain pipeline delete from t1 where a = 2;
ERROR 1105 (HY000): TokioError. Code: 1068, Text = Cannot join handle from context's runtime, cause: task 362 panicked.

### How to Reproduce?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by JackTan25**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR
explain pipeline delete/update will cause panic.

- Closes #12597

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12596)

### CODE REVIEW

**Comment by BohuTANG** [comment]
What's the bug is? or we is it possible to add a logic test to explain the pipeline after fixed.

**Comment by SkyFan2002** [comment]
FYI, some DML will actually do some data mutation( and commit) in `execute2` instead of just building pipeline, so if you want to explain pipeline of DML, you should carefully check the `execute2`.  

Or for safety, suggest returning an unimplemented error for now , and implementing them in later pr.

**Comment by JackTan25** [comment]
> FYI, some DML will actually do some data mutation( and commit) in `execute2` instead of just building pipeline, so if you want to explain pipeline of DML, you should carefully check the `execute2`.
> 
> Or for safety, suggest returning an unimplemented error for now , and implementing them in later pr.

for this pr, I just give a flag for interpreter, and the old logic is the same

**Comment by SkyFan2002** [comment]
> > FYI, some DML will actually do some data mutation( and commit) in `execute2` instead of just building pipeline, so if you want to explain pipeline of DML, you should carefully check the `execute2`.
> > Or for safety, suggest returning an unimplemented error for now , and implementing them in later pr.
> 
> for this pr, I just give a flag for interpreter, and the old logic is the same. cc @SkyFan2002

I mean, if you call explain pipeline delete, some mutation may be really committed. https://github.com/datafuselabs/databend/blob/0c5dd54b47a2ab94e213d9dd517d7f52d3598071/src/query/storages/fuse/src/operations/delete.rs#L105

**Comment by JackTan25** [comment]
> > > FYI, some DML will actually do some data mutation( and commit) in `execute2` instead of just building pipeline, so if you want to explain pipeline of DML, you should carefully check the `execute2`.
> > > Or for safety, suggest returning an unimplemented error for now , and implementing them in later pr.
> > 
> > 
> > for this pr, I just give a flag for interpreter, and the old logic is the same. cc @SkyFan2002
> 
> I mean, if you call explain pipeline delete, some mutation may be really committed.
> 
> https://github.com/datafuselabs/databend/blob/0c5dd54b47a2ab94e213d9dd517d7f52d3598071/src/query/storages/fuse/src/operations/delete.rs#L105

yes, we need to track this, thanks.

File: src/query/service/src/interpreters/interpreter_delete.rs [hunk 1]
```diff
@@ -78,12 +78,21 @@ use crate::stream::PullingExecutorStream;
 pub struct DeleteInterpreter {
     ctx: Arc<QueryContext>,
     plan: DeletePlan,
+    explain_pipeline: bool,
```
Comment (zhyass) [inline src/query/service/src/interpreters/interpreter_delete.rs:81]: use `enabled_table_lock`

File: src/query/service/src/interpreters/interpreter_delete.rs [hunk 2]
```diff
@@ -106,12 +115,17 @@ impl Interpreter for DeleteInterpreter {
...
         let tbl = self.ctx.get_table(catalog_name, db_name, tbl_name).await?;
         let table_info = tbl.get_table_info().clone();
-
-        // Add table lock heartbeat.
-        let handler = TableLockHandlerWrapper::instance(self.ctx.clone());
-        let mut heartbeat = handler
-            .try_lock(self.ctx.clone(), table_info.clone())
-            .await?;
+        let explain_pipeline = self.explain_pipeline;
+        let mut heartbeat = None;
```
Comment (zhyass) [inline src/query/service/src/interpreters/interpreter_delete.rs:119]: if `enabled_table_lock` is false, create a `DummyTableLock`.

File: src/query/service/src/interpreters/interpreter_delete.rs [hunk 3]
```diff
@@ -251,13 +265,18 @@ impl Interpreter for DeleteInterpreter {
...
         }
-
+        let explain_pipeline = self.explain_pipeline;
         if build_res.main_pipeline.is_empty() {
-            heartbeat.shutdown().await?;
+            if !explain_pipeline {
+                heartbeat.unwrap().shutdown().await?;
+            }
         } else {
             build_res.main_pipeline.set_on_finished(move |may_error| {
```
Comment (zhyass) [inline src/query/service/src/interpreters/interpreter_delete.rs:274]:
```rust
        if build_res.main_pipeline.is_empty() {
            heartbeat.shutdown().await?;
        } else {
            if enable_table_lock {
                build_res.main_pipeline.set_on_finished(move |may_error| {
                    // shutdown table lock heartbeat.
                    GlobalIORuntime::instance().block_on(async move { heartbeat.shutdown().await })?;
                    match may_error {
                        None => Ok(()),
                        Some(error_code) => Err(error_code.clone()),
                    }
                });
            }
        }
```

**Comment by JackTan25** [comment]
I check update/replace/insert/recluster, it's safe. delete will do_truncate, so I add some logics. For compact, if compact block, it's safe, so hook_compact is also safe. But for compact_segment, it's not so, so I add some logic. cc @dantengsky @zhyass @SkyFan2002

**Comment by SkyFan2002** [comment]
Another truncate:
https://github.com/datafuselabs/databend/blob/0c5dd54b47a2ab94e213d9dd517d7f52d3598071/src/query/storages/fuse/src/operations/delete.rs#L128

**Comment by JackTan25** [comment]
> Another truncate:
> 
> https://github.com/datafuselabs/databend/blob/0c5dd54b47a2ab94e213d9dd517d7f52d3598071/src/query/storages/fuse/src/operations/delete.rs#L128

fixed. Thanks @SkyFan2002 , you are more careful.

**Review by SkyFan2002 — APPROVED** [review]

File: src/query/service/src/interpreters/interpreter_factory.rs [hunk 4]
```diff
@@ -53,17 +53,25 @@ pub struct InterpreterFactory;
 /// Such as: Plan::Query -> InterpreterSelectV2
 impl InterpreterFactory {
     #[async_backtrace::framed]
-    pub async fn get(ctx: Arc<QueryContext>, plan: &Plan) -> Result<InterpreterPtr> {
+    pub async fn get(
+        ctx: Arc<QueryContext>,
+        plan: &Plan,
+        is_explain_pipeline: bool,
```
Comment (sundy-li) [inline src/query/service/src/interpreters/interpreter_factory.rs:59]: `is_explain_pipeline` could be derived from plan.

```
let is_explain_pipeline = match plan ...
```

Comment (JackTan25) [reply src/query/service/src/interpreters/interpreter_factory.rs:59]: that's better, thanks

Comment (JackTan25) [reply src/query/service/src/interpreters/interpreter_factory.rs:59]: we can't use this method, when we arrive here, it's the real plan like delete/update, not explain_pipeline plan.

**Comment by dantengsky** [comment]
@JackTan25 

really appreciate it. 

although the way this PR addresses the issue should work, it looks a little bit hacky.

could we instead, make the `update / delete` etc. follow the principle that,  in method `Interpreter::execute2`, no "side effects" but only the pipeline should be constructed?

e.g. the table-locking related actions should only be taken in `Interpreter::execute` (we may need to extend the `Interpreter` trait or introduce other mechanisms, to express that during execution, something should be locked)

I think that is the original idea/principle that impls of `Interpreter::execute2` should follow.  @zhang2014 need your opinion, please.

------
besides

seems that during explaining  `optimize table ...`, the following action should be prohibited as well

https://github.com/datafuselabs/databend/blob/744cf3ae5e6cb5efebe38acf39fa9d69be1efd17/src/query/service/src/interpreters/interpreter_table_optimize.rs#L115-L121

**Comment by dantengsky** [comment]
> could we instead, make the update / delete etc. follow the principle that, in method Interpreter::execute2, no "side effects" but only the pipeline should be constructed?

that would be harder, but if necessary, we can do the refacts step by step.

**Comment by JackTan25** [comment]
stop merge this pr first. Maybe this pr will cause some errors unexpectedly, a little hacky.

**Pull request closed without merging**